src/neo/SmartContract/ApplicationEngine.Runtime.cs

```diff
@@ -133,7 +133,8 @@ protected internal ulong GetRandom()
             // Block related
             for (int count = 0; count < ProtocolSettings.ValidatorsCount; index--, count++)
             {
-                System.Array.Copy(NativeContract.Ledger.GetBlockHash(Snapshot, index).ToArray(), 0, hash, offset + (UInt256.Length * count), UInt256.Length);
+                var blockHash = count == 0 ? PersistingBlock.Hash : NativeContract.Ledger.GetBlockHash(Snapshot, index);
```

Comment:
PersistingBlock.Hash is decidable to the primary, this makes the random number not unique.